### ISSUE TITLE
[Windows][MinGW 4.9.1] Cannot compile: "invalid option -- M"

### ISSUE THREAD

**Issue by arkdemon** [daokoder/dao#395]
Hello!
I tried compiling DAO by using this command: mingw32-make -f Makefile.daomake mingw" but it does not work. I get the following error: "invalid option -- M" (which refers to "-DMINGW" of line 149 of file dao-master\tools\daomake\bootstrap\Makefile)

I wanted to know which files I have to include to build DAO. Or is it possible to use CMake? (I may be able to generate one for Window MinGW when I have the time, but it might be in several weeks)

I really like the syntax of the script language and it also seems to have all the features I need as well as looking pretty fast ^^

PS: As I use Qt Creator as IDE, I tried including all the files in the folder kernel and compiling it. It compiles fine but I do not know how to run a script.I also found a lot of warnings in the code about unused variables, which can be removed easily.

**Comment by Night-walker** [comment]
I guess you should try an earlier MinGW, like 4.8 or so -- it has always been buggy. MinGW 4.8.2 I have works fine with the exact same command. CMake is currently not supported, so building with daomake is basically the only option.

**Comment by arkdemon** [comment]
Thank you :)
I thought it might be the problem but I was too lazy to reinstall MinGW...

**Comment by arkdemon** [comment]
I still get the error with MinGW 4.8.1-4 from http://sourceforge.net/projects/mingw/files/ (mingw-get-setup.exe)

**Comment by Night-walker** [comment]
OK, can you post full make log? How does expanded command line which triggers the error look like? If you're sure it's because of "-DMINGW", you could (as a temporary workaround) try removing this switch and adding `#define MINGW` at the top of `kernel/dao.h`.

**Comment by arkdemon** [comment]
Hi! sorry but I can't do it right now. I'll send it to you later (normally this weekend)

**Comment by arkdemon** [comment]
Hi! I fixed the issue by 'manually' adding all the files in a CMake project. Here is a small CMakeLists project (without the demos) (it should also be portable) (I only tested it on my computer so it might be buggy/not work):

CMakeLists.txt:
cmake_minimum_required(VERSION 2.8)
project(dao)

option(DAO_BUILD_TYPE "Set dao's build type: STATIC or SHARED" STATIC)
option(DAO_ENABLE_THREADS "Enable threads" ON)
option(DAO_ENABLE_NUMARRAY "Enable numarrays" ON)

#binary and library directory
SET(CMAKE_BINARY_DIR ${CMAKE_SOURCE_DIR})
SET(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/lib)
SET(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/lib)
SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/bin)

#Definitions - I don't know all of them
if(DAO_ENABLE_THREADS)
    add_definitions(-DDAO_WITH_THREAD)
endif()

if(DAO_ENABLE_NUMARRAY)
    add_definitions(-DDAO_WITH_NUMARRAY)
endif()

# set sources: .c and .h (.h are optional - I put them there so that they show up in qt creator)
FILE( GLOB_RECURSE DAO_SOURCES  "kernel/*.c" "modules/*.c" "kernel/*.h" "modules/*.h")

#create library with DAO_SOURCES as source and build type DAO_BUILD_TYPE
add_library(${PROJECT_NAME} ${DAO_BUILD_TYPE} ${DAO_SOURCES})

#include directories (for the dao lib)
INCLUDE_DIRECTORIES("${CMAKE_SOURCE_DIR}/kernel")
INCLUDE_DIRECTORIES("${CMAKE_SOURCE_DIR}/modules")

**Comment by arkdemon** [comment]
Here is the log if you like:
cd ./tools/daomake/bootstrap && mingw32-make mingw PLAT=mingw
mingw32-make[1]: Entering directory 'c:/Users/Me/Downloads/dao-master/tools/daomake/bootstrap'
c:/Qt/Tools/mingw491_32/bin/mingw32-make DEBUG= SYSCFLAGS="-DWIN32 -DMINGW" SYSLIBS="-lwinmm" EXESUFFIX=".exe"
c:\Qt\Tools\mingw491_32\bin\mingw32-make: invalid option -- M
Usage: mingw32-make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
  --eval=STRING               Evaluate STRING as a makefile statement.
  -f FILE, --file=FILE, --makefile=FILE
                              Read FILE as a makefile.
  -h, --help                  Print this message and exit.
  -i, --ignore-errors         Ignore errors from recipes.
  -I DIRECTORY, --include-dir=DIRECTORY
                              Search DIRECTORY for included makefiles.
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
  -k, --keep-going            Keep going when some targets can't be made.
  -l [N], --load-average[=N], --max-load[=N]
                              Don't start multiple jobs unless load is below N.
  -L, --check-symlink-times   Use the latest mtime between symlinks and target.
  -n, --just-print, --dry-run, --recon
                              Don't actually run any recipe; just print them.
  -o FILE, --old-file=FILE, --assume-old=FILE
                              Consider FILE to be very old and don't remake it.
  -O[TYPE], --output-sync[=TYPE]
                              Synchronize output of parallel jobs by TYPE.
  -p, --print-data-base       Print make's internal database.
  -q, --question              Run no recipe; exit status says if up to date.
  -r, --no-builtin-rules      Disable the built-in implicit rules.
  -R, --no-builtin-variables  Disable the built-in variable settings.
  -s, --silent, --quiet       Don't echo recipes.
  -S, --no-keep-going, --stop
                              Turns off -k.
  -t, --touch                 Touch targets instead of remaking them.
  --trace                     Print tracing information.
  -v, --version               Print the version number of make and exit.
  -w, --print-directory       Print the current directory.
  --no-print-directory        Turn off -w, even if it was turned on implicitly.
  -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
                              Consider FILE to be infinitely new.
  --warn-undefined-variables  Warn when an undefined variable is referenced.

This program built for i686-w64-mingw32
Report bugs to <bug-make@gnu.org>
makefile:149: recipe for target 'mingw' failed
mingw32-make[1]: *** [mingw] Error 2
mingw32-make[1]: Leaving directory 'c:/Users/Me/Downloads/dao-master/tools/daomake/bootstrap'
Makefile.daomake:45: recipe for target 'mingw' failed
mingw32-make: *** [mingw] Error 2

**Comment by Night-walker** [comment]
I still don't see what's the problem with mingw32-make, mine does not report invalid flag given the same command line. The variant without -DMINGW clearly fails because you don't have CC variable pointed to the name of the compiler you're using -- "set CC=gcc" should do the trick.

**Comment by arkdemon** [comment]
Adding CC=gcc made it compile! Thank you!

But, I also got the error: daoPlatforms-mingw.o:daoPlatforms.c:(.text+0x136): undefined reference to `_imp__timeGetTime@0'
that I fix by transforming the #ifdef WIN32 to #ifded MSVC

and to compile the module help I get this error:
/usr/bin/sh: -c: line 0: syntax error near unexpected token `en'
/usr/bin/sh: -c: line 0: `C:/Users/Me/Downloads/dao-master/bin/dao --path=\C:/Users/Me/Downloads/dao-master/modules/help/\ -e \load help; help::set_language(en); help::set_font(monospace); help::export(,./doc/html/en)\'

**Comment by Night-walker** [comment]
> I also compiled dao-modules but regex does not work.

The reason of that is the incomplete makefile; I wasn't sure how to write makefile.dao in this case (the module was written by me), so I postponed the issue. If you want to use it, the instructions in README.md should help.

**Comment by arkdemon** [comment]
Actually it was only to point it out: I don't use it. 
I am quite impressed by the huge number of modules though x). There seems to be just about everything and it seems easy to add custom modules for diffeent applications.
(I don't know if it pretentious to say it - if so, it is not my intention): Great job! The language is really fantastic and the 'standard 'library' too!

PS: I don't know if it is the right place to ask but:
1. in which module is std.listMeth? or an equivalent? 
2. What does DaoCXX do? Does it compile C++ code? or converts c++ code to the dao language? or something else?
3. how can I compile DaoCXX? 
4. and DaoJIT? 

Thank you!

**Comment by Night-walker** [comment]
> 1. in which module is std.listMeth? or an equivalent? (or is it deprecated?)

I guess there isn't one anymore... Documentation is a work in progress, so some things can only be known from the sources for the moment. Or from us adepts of arcane Dao knowledge :)

> 2. What does DaoCXX do? Does it compile C++ code? or converts c++ code to the dao language? or something else?

If I remember correctly it is related to LLVM-backed JIT compilation provided by DaoJIT.

> 3. how can I compile DaoCXX?

You need Clang ("set CC=clang"), then DaoCXX should be included in the build process. Haven't tried to build it for a long time though.

> 4. and DaoJIT?

Should be the same as with DaoCXX.

**Comment by arkdemon** [comment]
Thank you for your mystical advice, wise adept.
It has enlightened my poor soul :p

**Comment by dumblob** [comment]
>EDIT: is it normal that 'load os' in script crashes?

No, it's not (at least it works for me). Could you please put here a valgrind or gdb stack trace or any similar report?

**Comment by arkdemon** [comment]
How do I create a gdb stack trace?

**Comment by Night-walker** [comment]
Don't bother. There's a piece of code there which is running amok. Will try to sort it out.

**Comment by arkdemon** [comment]
I have 2 logs, corresponding to the same thing. (one is not verbose I think, the other is)
(1 week duration)
Log 1: http://pastebin.com/n0ab3t52
Log 2 (beware it is very long): http://pastebin.com/usGcXvN6

**Comment by Night-walker** [comment]
It's an API issue caused by the silly way things get linked on Windows. I temporarily disabled `os::PipeStream`; when @daokoder shows up, he'll be able to fix the API and get things back to normal.

**Comment by arkdemon** [comment]
Ok, thank you.
